rbreslow/zs
# Add map voting system

## Overview

The following changes were made to support the addition of a map voting system:

- Add grid system for laying out VGUI
- Add map vote VGUI elements
- Add migration to create `maps` table for storing statistics
- Add system to sync maps with database on initialization and pool maps at the end of the round for a vote
- Add 576x324 sized images for each map as well as localized map name strings
- Add database queue and mechanism for recovering dead connections

## Testing Instructions

First, update FastDL and run database migrations.

```bash
$ ./scripts/update
. . .
SQL: SELECT COUNT(*) as c from pg_class where relname = 'lapis_migrations'
SQL: SELECT * from "lapis_migrations"
Migrating: 1598730685
SQL: CREATE TABLE "maps" (
  "id" character varying(255) NOT NULL PRIMARY KEY,
  "last_played" timestamp NOT NULL DEFAULT '-infinity',
  "installed" boolean NOT NULL DEFAULT TRUE
);
SQL: INSERT INTO "lapis_migrations" ("name") VALUES ('1598730685') RETURNING "name"
Ran 1 migration
```

Bring up SRCDS, and then in the server console, advance the current round to the round limit and end the round to trigger a map vote.

```bash
$ ./scripts/console
lua_run GAMEMODE.CurrentRound = 3
> GAMEMODE.CurrentRound = 3...
lua_run GAMEMODE:EndRound(TEAM_ZOMBIE)
> GAMEMODE:EndRound(TEAM_ZOMBIE)...
```

Vote for a map, and see that the server takes you there.

![image](https://user-images.githubusercontent.com/1774125/92340638-2fb81380-f089-11ea-8c5b-f24828bbd30f.png)
